inmem_endpoint.go

```diff
@@ -52,10 +52,10 @@ func (i *InmemSink) DisplayMetrics(resp http.ResponseWriter, req *http.Request)
 		return nil, fmt.Errorf("no metric intervals have been initialized yet")
 	case n == 1:
 		// Show the current interval if it's all we have
-		interval = i.intervals[0]
+		interval = data[0]
 	default:
 		// Show the most recent finished interval if we have one
-		interval = i.intervals[n-2]
+		interval = data[n-2]
 	}
 
```

Comment:
I believe you need to lock the interval while reading from it too, e.g.

```suggestion
	interval.RLock()
	defer interval.RUnlock()

```

Reply:
In the current version of the code, this is a deep copy. See `Data()`

Reply:
`Data()` only deep copies the last element `n-1`in https://github.com/armon/go-metrics/blob/master/inmem.go#L241-L264 - but we may be reading element at  `n-2`.  I can think of possible scenarios where `SetGaugesWithLabels` for example gets `n-1` to manipulate it while another goroutine appends to interval array, resulting into a race of modification of `SetGaugesWithLabels` and read in this function.

Also, I see that `dumpStats()` still locks elements when iterating through elements returned from `Data()` in https://github.com/armon/go-metrics/blob/master/inmem_signal.go#L77-L100 .

Unless we have good coverage of tests that pass with `-race`, I'd suggest being defensive here.